<a href="https://colab.research.google.com/github/taehi5/Esaa_YB/blob/4%EC%A3%BC%EC%B0%A8-%EA%B3%BC%EC%A0%9C/ESAA_YB_Week4_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

`ch2 . 사이킷런으로 시작하는 머신러닝`

# 01. 사이킷런 소개와 특징
- 머신러닝 라이브러리 중 가장 많이 사용됨
- 가장 파이썬스러운 API제공
- 머신러닝을 위한 매우 다양한 알고리즘과 개발을 위한 편리한 프레임워크&API 제공
- 오랜 기간 실전 환경에서 검증, 매우 많은 환경에서 사용되는 성숙한 라이브러리

In [2]:
pip install scikit-learn==1.0.2

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.7/6.7 MB 66.4 MB/s eta 0:00:00
  Installing build dependencies ... done
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Getting requirements to build wheel ... error
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.


In [3]:
import sklearn

print(sklearn.__version__)

1.6.1


# 02. 첫 번째 머신러닝 만들어 보기 - 붓꽃 품종 예측하기
▶ 붓꽃 데이터 세트 : 꽃잎의 길이, 너비, 꽃받침의 길이와 너비 피처
목표 : 붓꽃 데이터 세트로 붓꽃의 품종 분류  



 ### 분류(Classification)
 -  대표적인 지도 학습 방법 중 하나
> 지도학습 : 학습을 위한 다양한 피처와 레이블 데이터로 모델을 학습, 별도의 테스트 데이터 세트에서 미지의 레이블을 예측
>- 명확한 정답이 주어진 데이터를 먼저 학습한 뒤 미지의 정답을 예측하는 방식
> 이때 주어진 데이터 셋을 학습 데이터와 테스트 데이터로 나눔




### 모듈 임포트
- 사이킷런 패키지 내의 모듈명은 sklearn으로 시작
- **sklearn.datasets**: 사이킷런에서 자체적으로 제공하는 데이터 셋 생성 모듈의 모임
- **sklearn.tree**: 트리 기반 ml알고리즘을 구현한 클래스의 모임
- **sklearn.model_selection** : 학습 데이터와 검증 데이터, 예측 데이터로 데이터를 분리하거나 최적의 하이퍼 파라미터로 평가하기 위한 모듈의 모임
> 하이퍼 파라미터 : 머신러닝 알고리즘별로 최적의 학습을 위해 직접 입력하는 파라미터를 통칭, 머신러닝 알고리즘의 성능 튜닝 가능


In [4]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [5]:
# load_iris()를 붓꽃 데이터 세트를 로딩, df로 변환
import pandas as pd

iris = load_iris()

iris_data = iris.data

iris_label = iris.target
print('iris target값:',iris_label)
print('iris target명:',iris.target_names)

# 붓꽃 데이터 세트를 자세히 보기 위해 DataFrame으로 변환
iris_df = pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df['label'] = iris.target
iris_df.head(3)

iris target값: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target명: ['setosa' 'versicolor' 'virginica']


,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),label
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0


### 학습 데이터와 테스트 데이터 분리
- train_test_split 사용

In [6]:
X_train,X_test,y_train,y_test = train_test_split(iris_data,iris_label,test_size=0.2,random_state=11)

파라미터 설명
- iris_data : 피처 데이터 세트
- iris_label : 레이블 데이터 세트
- test_size = 0.2 : 전체 데이터 세트 중 테스트 데이터 세트의 비율
- random_state : 호출할 때마다 같은 학습/테스트 용 데이터 세트를 생성하기 위해 주어지는 난수 발생값

### 의사 결정 트리를 이용해 학습/예측 수행
- 사이킷런의 의사 결정 트리 클래스 사용

In [7]:
dt_clf = DecisionTreeClassifier(random_state=11)

dt_clf.fit(X_train,y_train)

DecisionTreeClassifier(random_state=11)

### 예측 수행
- 학습 데이터가 아닌 다른 데이터(test data)

In [8]:
pred = dt_clf.predict(X_test)

### 의사 결정 트리 기반의 DecisionTreeClassifier의 예측 성능 평가
- 정확도: 예측 결과가 실제 레이블 값과 얼마나 정확하게 맞는지를 평가하는 지표
- accuracy_score() 함수 이용  

In [9]:
from sklearn.metrics import accuracy_score
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test,pred)))

예측 정확도: 0.9333


붓꽃 데이터 세트로 분류를 예측한 프로세스
1. 데이터 세트 분리 : 학습/테스트로 분리
2. 모델 학습: 학습 데이터 기반으로 ml 알고리즘 적용
3. 예측 수행: ml 모델을 이용해 데이터의 분류 예측
4. 평가: 예측값과 실제 결괏값을 비교

# 03. 사이킷런의 기반 프레임워크 익히기


##  1) Estimator 이해 및 fit(), predict()메서드
 1. fit()
 : ml 모델 학습을 위함
 2. predict()
 : 학습된 모델의 예측을 위함
 3. Estimator : Classifier + Regressor
 - 지도학습의 모든 알고리즘을 구현한 클래스 통칭

- evaluation, 하이퍼 파라미터 튜닝을 지원하는 클래스가 estimator를 인자로 받음

## 2) 사이킷런의 주요 모듈


| 분류 | 모듈명 | 설명 |
|---|---|---|
| 예제 데이터 | `sklearn.datasets` | 사이킷런에 내장되어 예제로 제공하는 데이터 세트 |
| 피처 처리 | `sklearn.preprocessing` | 데이터 전처리에 필요한 다양한 가공 기능 제공(문자열을 숫자형 코드 값으로 인코딩, 정규화, 스케일링 등) |
| 피처 처리 | `sklearn.feature_selection` | 알고리즘에 큰 영향을 미치는 피처를 우선순위대로 셀렉션 작업을 수행하는 다양한 기능 제공 |
| 피처 처리 | `sklearn.feature_extraction` | 텍스트 데이터나 이미지 데이터의 벡터화된 피처를 추출하는 데 사용됨.<br>예를 들어 텍스트 데이터에서 Count Vectorizer나 Tf-idf Vectorizer 등을 생성하는 기능 제공.<br>텍스트 데이터의 피처 추출은 `sklearn.feature_extraction.text` 모듈이, 이미지 데이터의 피처 추출은 `sklearn.feature_extraction.image` 모듈에 지원 API가 있음. |
| 피처 처리 & 차원 축소 | `sklearn.decomposition` | 차원 축소와 관련한 알고리즘을 지원하는 모듈. PCA, NMF, Truncated SVD 등을 통해 차원 축소 기능을 수행할 수 있음. |
| 데이터 세트 분리 & 파라미터 튜닝 | `sklearn.model_selection` | 교차 검증을 위한 학습용/테스트용 분리, 그리드 서치(Grid Search)로 최적 파라미터 추출 등의 API 제공 |
| 평가 | `sklearn.metrics` | 분류, 회귀, 클러스터링, 페어와이즈(Pairwise)에 대한 다양한 성능 측정 방법 제공.<br>Accuracy, Precision, Recall, ROC-AUC, RMSE 등 제공 |
| ML 알고리즘 | `sklearn.ensemble` | 앙상블 알고리즘 제공.<br>랜덤 포레스트, 에이다 부스트, 그래디언트 부스팅 등을 제공 |
| ML 알고리즘 | `sklearn.linear_model` | 주로 선형 회귀, 릿지(Ridge), 라쏘(Lasso) 및 로지스틱 회귀 등 회귀 관련 알고리즘 지원. 또한 SGD(Stochastic Gradient Descent) 관련 알고리즘도 제공 |
| ML 알고리즘 | `sklearn.naive_bayes` | 나이브 베이즈 알고리즘 제공. 가우시안 NB, 다항 분포 NB 등. |
| ML 알고리즘 | `sklearn.neighbors` | 최근접 이웃 알고리즘 제공. K-NN 등. |
| ML 알고리즘 | `sklearn.svm` | 서포트 벡터 머신 알고리즘 제공 |
| ML 알고리즘 | `sklearn.tree` | 의사 결정 트리 알고리즘 제공 |
| ML 알고리즘 | `sklearn.cluster` | 비지도 클러스터링 알고리즘 제공.<br>K-평균, 계층형, DBSCAN 등. |
| 유틸리티 | `sklearn.pipeline` | 피처 처리 등의 변환과 ML 알고리즘 학습, 예측 등을 함께 묶어서 실행할 수 있는 유틸리티 제공 |

## 내장된 예제 데이터 세트
### 분류·회귀 연습용 예제 데이터

| API 명 | 설명 |
|---|---|
| `datasets.load_boston()` | 회귀용. 미국 보스턴의 집 피처들과 가격에 대한 데이터 세트 |
| `datasets.load_breast_cancer()` | 분류용. 위스콘신 유방암 피처들과 악성·양성 레이블 데이터 세트 |
| `datasets.load_diabetes()` | 회귀용. 당뇨 데이터 세트 |
| `datasets.load_digits()` | 분류용. 0~9까지 숫자의 이미지 픽셀 데이터 세트 |
| `datasets.load_iris()` | 분류용. 붓꽃에 대한 피처를 가진 데이터 세트 |

### `fetch` 계열 데이터

`fetch` 계열은 데이터 크기가 커서 패키지에 처음부터 저장되어 있지 않습니다. 인터넷에서 내려받아 홈 디렉터리 아래의 `scikit_learn_data` 서브 디렉터리에 저장한 후, 이후에는 저장된 데이터를 불러옵니다.

따라서 **최초 사용 시에는 인터넷 연결이 필요합니다.**

| API 명 | 설명 |
|---|---|
| `fetch_covtype()` | 회귀 분석용 토지 조사 자료 |
| `fetch_20newsgroups()` | 뉴스 그룹 텍스트 자료 |
| `fetch_olivetti_faces()` | 얼굴 이미지 자료 |
| `fetch_lfw_people()` | 얼굴 이미지 자료 |
| `fetch_lfw_pairs()` | 얼굴 이미지 자료 |
| `fetch_rcv1()` | 로이터 뉴스 말뭉치 |
| `fetch_mldata()` | ML 웹사이트에서 다운로드 |


# 분류와 클러스터링을 위한 표본 데이터 생성기

| API 명 | 설명 |
|---|---|
| `datasets.make_classification()` | 분류를 위한 데이터 세트를 만듭니다. 특히 높은 상관도, 불필요한 속성 등의 노이즈 효과를 위한 데이터를 무작위로 생성해 줍니다. |
| `datasets.make_blobs()` | 클러스터링을 위한 데이터 세트를 무작위로 생성해 줍니다. 군집 지정 개수에 따라 여러 가지 클러스터링을 위한 데이터 세트를 쉽게 만들어 줍니다. |

표본 데이터 생성기는 이 밖에도 많으며, 위의 2개 정도로도 여러 가지 사례에 사용할 수 있어 여기에서 소개를 마칩니다.

## 예제 데이터 세트의 구성

분류나 회귀를 위한 연습용 예제 데이터가 어떻게 구성돼 있는지 좀 더 살펴보겠습니다. 사이킷런에 내장된 이 데이터 세트는 일반적으로 딕셔너리 형태로 돼 있습니다.

키는 보통 `data`, `target`, `target_names`, `feature_names`, `DESCR`로 구성돼 있습니다. 개별 키가 가리키는 데이터 세트의 의미는 다음과 같습니다.

- `data`는 피처의 데이터 세트를 가리킵니다.
- `target`은 분류 시 레이블 값, 회귀일 때는 숫자 결과값 데이터 세트입니다.
- `target_names`는 개별 레이블의 이름을 나타냅니다.
- `feature_names`는 피처의 이름을 나타냅니다.
- `DESCR`은 데이터 세트에 대한 설명과 각 피처의 설명을 나타냅니다.


### 피처의 데이터 값 반환
: 내장 데이터 세트 API 호출 후 key 지정


In [10]:
from sklearn.datasets import load_iris

iris_data = load_iris()
print(type(iris_data))

<class 'sklearn.utils._bunch.Bunch'>


In [11]:
keys = iris_data.keys()
print('붓꽃 데이터 세트의 키들:', keys)

붓꽃 데이터 세트의 키들: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


데이터 키 : 피처들의 데이터 값
- 피처 데이터 값을 추출 : 데이터세트.data 이용




In [12]:
print('\n feature_names 의 type:',type(iris_data.feature_names))
print(' feature_names 의 shape:',len(iris_data.feature_names))
print(iris_data.feature_names)

print('\n target_names의 type:',type(iris_data.target_names))
print(' target_names 의 shape:',len(iris_data.target_names))
print(iris_data.target_names)

print('\n data의 type:',type(iris_data.data))
print(' data의 shape:',iris_data.data.shape)
print(iris_data['data'])

print('\n target의 type:',type(iris_data.target))
print('target의 shape:',iris_data.target.shape)
print(iris_data.target)


 feature_names 의 type: <class 'list'>
 feature_names 의 shape: 4
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

 target_names의 type: <class 'numpy.ndarray'>
 target_names 의 shape: 3
['setosa' 'versicolor' 'virginica']

 data의 type: <class 'numpy.ndarray'>
 data의 shape: (150, 4)
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]


# 04. Model Selection 모듈 소개
데이터 세트 분리, 교차 검증 , 하이퍼 파라미터 튜닝 등을 위한 다양한 함수와 클래스 제공

## train_test_split
학습/테스트 데이터 세트 분리

In [13]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

iris = load_iris()
dt_clf = DecisionTreeClassifier()
train_data = iris.data
train_label = iris.target
dt_clf.fit(train_data, train_label)

pred = dt_clf.predict(train_data)
print('예측 정확도:',accuracy_score(train_label,pred))

예측 정확도: 1.0


분리 과정
1. train_test_split 로드
2. 파라미터로 레이블 데이터 세트 입력
- test_size: 전체 중 테스트 크기를 얼마로 할 것인지 결정
- train_size:전체 중 학습용 크기를 얼마로 할 것인지 결정
- shuffle: 데이터 분리 전에 섞을지 결정
- random_state : 동일한 세트를 생성하기 위한 난수 값

In [14]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

dt_clf = DecisionTreeClassifier()
iris_data = load_iris()

X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size=0.3,random_state=121)


In [15]:
dt_clf.fit(X_train,y_train)
pred = dt_clf.predict(X_test)
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test,pred)))

예측 정확도: 0.9556


## 교차 검증
> 과적합: 모델이 학습 데이터에만 과도하게 최적화, 실제 예측을 다른 데이터로 수행할 경우에는 예측 성능이 과도하게 떨어지는 것


> ▶해결 방법: 교차 검증


교차 검증 : 별도의 여러 세트로 구성된 학습 데이터 세트와 검증 데이터 세트에서 학습과 평가를 수행

### K-fold 교차검증
: K개의 데이터 폴드 세트를 만들어서 K번 만큼 각 폴드 세트에 학습과 검증 평가를 반복적으로 수행하는 방법
> 예) 5폴드 교차 검증
> 1. 5개의 폴드로 된 데이터 세트를 학습과 검증을 위한 데이터 세트로 변경하면서 5번 평가 수행
> 2. 5개의 평가를 평균한 결과를 가지고 예측 성능 평가

### KFold 클래스를 이용해 붓꽃 데이터 세트 교차 검증 및 예측 정확도 확인

In [16]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import numpy as np

iris = load_iris()
features = iris.data
label = iris.target
dt_clf = DecisionTreeClassifier(random_state = 156)


kfold = KFold(n_splits=5)
cv_accuracy=[]
print('붓꽃 데이터 세트 크기:',features.shape[0])

붓꽃 데이터 세트 크기: 150


생성된 KFold 객체의 split을 호출해 전체 붓꽃 데이터를 5개로 분리

In [17]:
n_iter = 0

for train_index, test_index in kfold.split(features):
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = label[train_index], label[test_index]

    dt_clf.fit(X_train,y_train)
    pred = dt_clf.predict(X_test)
    n_iter +=1

    accuracy = np.round(accuracy_score(y_test,pred),4)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    print('\n#{0} 교차 검증 정확도 :{1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'.format(n_iter,accuracy,train_size,test_size))
    print('#{0} 검증 세트 인덱스:{1}'.format(n_iter,test_index))
    cv_accuracy.append(accuracy)

print('\n## 평균 검증 정확도:',np.mean(cv_accuracy))


#1 교차 검증 정확도 :1.0, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#1 검증 세트 인덱스:[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]

#2 교차 검증 정확도 :0.9667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#2 검증 세트 인덱스:[30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]

#3 교차 검증 정확도 :0.8667, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#3 검증 세트 인덱스:[60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]

#4 교차 검증 정확도 :0.9333, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#4 검증 세트 인덱스:[ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]

#5 교차 검증 정확도 :0.7333, 학습 데이터 크기: 120, 검증 데이터 크기: 30
#5 검증 세트 인덱스:[120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

## 평균 검증 정확도: 0.9


### Stratified K폴드
: 불균형한 분포도를 가진 레이블 데이터 집합을 위한 K폴드 방식
> 불균형한 분포도를 가진 레이블 데이터 집합 : 특정 레이블 값이 특이하게 많거나 매우 적어서 값의 분포가 한쪽으로 치우치는 것



- 원본 데이터의 레이블 분포를 고려해 학습/검증 데이터에도 비슷한 비율이 유지되도록 분리

#### 붓꽃 데이터의 레이블 분포 확인

In [18]:
import pandas as pd

iris = load_iris()
iris_df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
iris_df['label']=iris.target
iris_df['label'].value_counts()


,count
label,
0,50
1,50
2,50


- 레이블 0, 1, 2는 각각 50개
- KFold로 3개 폴드를 만들고 학습/검증 레이블 분포 확인

In [19]:
kfold = KFold(n_splits=3)
n_iter =0
for train_index, test_index in kfold.split(iris_df):
    n_iter += 1
    label_train= iris_df['label'].iloc[train_index]
    label_test= iris_df['label'].iloc[test_index]
    print('## 교차 검증: {0}'.format(n_iter))
    print('학습 레이블 데이터 분포:\n', label_train.value_counts())
    print('검증 레이블 데이터 분포:\n', label_test.value_counts())


## 교차 검증: 1
학습 레이블 데이터 분포:
 label
1    50
2    50
Name: count, dtype: int64
검증 레이블 데이터 분포:
 label
0    50
Name: count, dtype: int64
## 교차 검증: 2
학습 레이블 데이터 분포:
 label
0    50
2    50
Name: count, dtype: int64
검증 레이블 데이터 분포:
 label
1    50
Name: count, dtype: int64
## 교차 검증: 3
학습 레이블 데이터 분포:
 label
0    50
1    50
Name: count, dtype: int64
검증 레이블 데이터 분포:
 label
2    50
Name: count, dtype: int64


- 붓꽃 데이터가 레이블 순서대로 정렬되어 있어 학습/검증 데이터의 레이블이 완전히 다르게 나뉨
- 학습에서 보지 못한 레이블만 검증 데이터에 들어가므로 제대로 예측할 수 없음

#### StratifiedKFold로 분포 유지
- split()에 피처 데이터와 **레이블 데이터도 반드시 전달**

In [20]:
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(n_splits=3)
n_iter=0
for train_index, test_index in skf.split(iris_df, iris_df['label']):
    n_iter += 1
    label_train= iris_df['label'].iloc[train_index]
    label_test= iris_df['label'].iloc[test_index]
    print('## 교차 검증: {0}'.format(n_iter))
    print('학습 레이블 데이터 분포:\n', label_train.value_counts())
    print('검증 레이블 데이터 분포:\n', label_test.value_counts())


## 교차 검증: 1
학습 레이블 데이터 분포:
 label
2    34
0    33
1    33
Name: count, dtype: int64
검증 레이블 데이터 분포:
 label
0    17
1    17
2    16
Name: count, dtype: int64
## 교차 검증: 2
학습 레이블 데이터 분포:
 label
1    34
0    33
2    33
Name: count, dtype: int64
검증 레이블 데이터 분포:
 label
0    17
2    17
1    16
Name: count, dtype: int64
## 교차 검증: 3
학습 레이블 데이터 분포:
 label
0    34
1    33
2    33
Name: count, dtype: int64
검증 레이블 데이터 분포:
 label
1    17
2    17
0    16
Name: count, dtype: int64


- 각 폴드의 학습 데이터는 100개, 검증 데이터는 50개
- 학습/검증 데이터 모두 레이블 0, 1, 2가 비슷한 비율로 포함됨

#### StratifiedKFold로 학습 및 정확도 확인

In [21]:
dt_clf = DecisionTreeClassifier(random_state=156)

skfold = StratifiedKFold(n_splits=3)
n_iter=0
cv_accuracy=[]

# StratifiedKFold의 split() 호출 시 반드시 레이블 데이터 세트도 추가 입력 필요
for train_index, test_index in skfold.split(features, label):
    # split()으로 반환된 인덱스를 이용해 학습용, 검증용 테스트 데이터 추출
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = label[train_index], label[test_index]
    # 학습 및 예측
    dt_clf.fit(X_train, y_train)
    pred = dt_clf.predict(X_test)

    # 반복 시마다 정확도 측정
    n_iter += 1
    accuracy = np.round(accuracy_score(y_test, pred), 4)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    print('\n#{0} 교차 검증 정확도 :{1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'
          .format(n_iter, accuracy, train_size, test_size))
    print('#{0} 검증 세트 인덱스:{1}'.format(n_iter, test_index))
    cv_accuracy.append(accuracy)

# 교차 검증별 정확도 및 평균 정확도 계산
print('\n## 교차 검증별 정확도:', np.round(cv_accuracy, 4))
print('## 평균 검증 정확도:', np.round(np.mean(cv_accuracy), 4))



#1 교차 검증 정확도 :0.98, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#1 검증 세트 인덱스:[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115]

#2 교차 검증 정확도 :0.94, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#2 검증 세트 인덱스:[ 17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  67
  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82 116 117 118
 119 120 121 122 123 124 125 126 127 128 129 130 131 132]

#3 교차 검증 정확도 :0.98, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#3 검증 세트 인덱스:[ 34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  83  84
  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 133 134 135
 136 137 138 139 140 141 142 143 144 145 146 147 148 149]

## 교차 검증별 정확도: [0.98 0.94 0.98]
## 평균 검증 정확도: 0.9667


- 교재 결과: 교차 검증별 정확도 0.98, 0.94, 0.98 / 평균 0.9667
- 분류에서는 레이블 비율을 유지하는 Stratified K폴드를 주로 사용
- 회귀의 타깃은 연속적인 숫자이므로 Stratified K폴드가 지원되지 않음

### 교차 검증을 보다 간편하게 - cross_val_score()
- 데이터 분할 → 학습 → 예측 → 평가 과정을 한 번에 수행
- estimator: 분류기 또는 회귀 모델
- X: 피처 데이터 / y: 레이블 데이터
- scoring: 성능 평가 지표 / cv: 교차 검증 폴드 수
- 폴드별 평가 점수를 배열로 반환, 평균을 구해 전체 성능 확인
- cv에 정수를 지정하면 분류는 StratifiedKFold, 회귀는 KFold 사용

In [22]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.datasets import load_iris

iris_data = load_iris()
dt_clf = DecisionTreeClassifier(random_state=156)

data = iris_data.data
label = iris_data.target

# 성능 지표는 정확도(accuracy), 교차 검증 세트는 3개
scores = cross_val_score(dt_clf, data, label, scoring='accuracy', cv=3)
print('교차 검증별 정확도:', np.round(scores, 4))
print('평균 검증 정확도:', np.round(np.mean(scores), 4))


교차 검증별 정확도: [0.98 0.94 0.98]
평균 검증 정확도: 0.9667


- 앞의 StratifiedKFold 예제와 동일한 검증 결과
- cross_validate(): 여러 평가 지표와 수행 시간 확인 가능, 학습 점수도 옵션으로 제공

### GridSearchCV - 교차 검증과 최적 하이퍼 파라미터 튜닝을 한 번에
- 하이퍼 파라미터: 모델의 학습 방식이나 복잡도를 조절하기 위해 지정하는 값
- 후보 값의 모든 조합에 교차 검증을 수행해 가장 좋은 조합을 찾음

In [23]:
grid_parameters = {'max_depth': [1, 2, 3],
                   'min_samples_split': [2, 3]
                  }


- max_depth 3개 × min_samples_split 2개 = 6개 조합
- cv=3이면 6 × 3 = 총 18회 학습/평가, refit에 따른 재학습은 별도
- 후보 조합이 많아질수록 수행 시간 증가

주요 파라미터
- estimator: 분류기, 회귀 모델, 파이프라인
- param_grid: 파라미터명과 후보 값 리스트로 구성된 딕셔너리
- scoring: 성능 평가 방법
- cv: 교차 검증 폴드 수
- refit: 기본값 True, 최적 파라미터로 전체 학습 데이터를 다시 학습

#### 붓꽃 데이터로 최적 파라미터 탐색

In [24]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# 데이터를 로딩하고 학습 데이터와 테스트 데이터 분리
iris_data = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target,                                               test_size=0.2, random_state=121)
dtree = DecisionTreeClassifier()

### 파라미터를 딕셔너리 형태로 설정
parameters = {'max_depth':[1, 2, 3], 'min_samples_split':[2, 3]}


In [25]:
import pandas as pd

# param_grid의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트 수행 설정.
### refit=True가 default임. True이면 가장 좋은 파라미터 설정으로 재학습시킴.
grid_dtree = GridSearchCV(dtree, param_grid=parameters, cv=3, refit=True)

# 붓꽃 학습 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습/평가.
grid_dtree.fit(X_train, y_train)

# GridSearchCV 결과를 추출해 DataFrame으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score',
           'split0_test_score', 'split1_test_score', 'split2_test_score']]


,params,mean_test_score,rank_test_score,split0_test_score,split1_test_score,split2_test_score
0,"{'max_depth': 1, 'min_samples_split': 2}",0.700000,5,0.700,0.7,0.70
1,"{'max_depth': 1, 'min_samples_split': 3}",0.700000,5,0.700,0.7,0.70
2,"{'max_depth': 2, 'min_samples_split': 2}",0.958333,3,0.925,1.0,0.95
3,"{'max_depth': 2, 'min_samples_split': 3}",0.958333,3,0.925,1.0,0.95
4,"{'max_depth': 3, 'min_samples_split': 2}",0.975000,1,0.975,1.0,0.95
5,"{'max_depth': 3, 'min_samples_split': 3}",0.975000,1,0.975,1.0,0.95


cv_results_ : 파라미터 조합별 교차 검증 결과
- params: 적용한 파라미터 조합
- rank_test_score: 성능 순위, 1이 가장 좋은 조합
- mean_test_score: 폴드별 검증 점수의 평균
- split0_test_score 등: 각 폴드의 검증 점수

최적 결과 확인
- best_params_: 가장 좋은 성능의 파라미터 조합
- best_score_: 해당 조합의 평균 교차 검증 점수

In [26]:
print('GridSearchCV 최적 파라미터:', grid_dtree.best_params_)
print('GridSearchCV 최고 정확도:{0:.4f}'.format(grid_dtree.best_score_))


GridSearchCV 최적 파라미터: {'max_depth': 3, 'min_samples_split': 2}
GridSearchCV 최고 정확도:0.9750


- 교재 결과: max_depth=3, min_samples_split=2 / 최고 검증 정확도 0.9750
- best_estimator_: refit=True일 때 최적 파라미터로 다시 학습된 모델
- 별도로 fit()할 필요 없이 테스트 데이터 예측 가능

In [27]:
# GridSearchCV의 refit으로 이미 학습된 estimator 반환
estimator = grid_dtree.best_estimator_

# GridSearchCV의 best_estimator_는 이미 최적 학습이 됐으므로 별도 학습이 필요 없음
pred = estimator.predict(X_test)
print('테스트 데이터 세트 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))


테스트 데이터 세트 정확도: 0.9667
